jena-core/src/main/java/org/apache/jena/util/FileManagerImpl.java

```diff
@@ -0,0 +1,566 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.jena.util;
+
+import java.io.* ;
+import java.util.* ;
+
+import org.apache.jena.atlas.logging.FmtLog;
+import org.apache.jena.rdf.model.Model ;
+import org.apache.jena.rdf.model.ModelFactory ;
+import org.apache.jena.shared.JenaException ;
+import org.apache.jena.shared.NotFoundException ;
+import org.apache.jena.shared.WrappedIOException ;
+import org.apache.jena.sys.JenaSystem ;
+import org.slf4j.Logger ;
+import org.slf4j.LoggerFactory ;
+
+/** FileManager
+ * 
+ * A FileManager provides access to named file-like resources by opening
+ * InputStreams to things in the filing system, by URL (http: and file:) and
+ * found by the classloader.  It can also load RDF data from such a system
+ * resource into an existing model or create a new (Memory-based) model.
+ * There is a global FileManager which provide uniform access to system
+ * resources: applications may also create specialised FileManagers.
+ * 
+ * A FileManager contains a list of location functions to try: the global
+ * FileManger has one {@link LocatorFile}, one {@link LocatorClassLoader} and
+ * one {@link LocatorURL}
+ * 
+ * Main operations:
+ *  * <ul>
+ * <li>loadModel, readModel : URI to model</li>
+ * <li>open, openNoMap : URI to input stream</li>
+ * <li>mapURI : map URI to another by {@link LocationMapper}</li> 
+ * </ul>
+ * 
+ * Utilities:
+ * <ul>
+ * <li>readWholeFileAsUTF8</li>
+ * <li>optional caching of models<li>
+ * </ul>
+ * 
+ * A FileManager works in conjunction with a LocationMapper.
+ * A {@link LocationMapper} is a set of alternative locations for system
+ * resources and a set of alternative prefix locations.  For example, a local
+ * copy of a common RDF dataset may be used whenever the usual URL is used by
+ * the application.
+ *
+ * The {@link LocatorFile} also supports the idea of "current directory".
+ * 
+ * @see LocationMapper
+ * @see FileUtils
+ */
+ 
+public class FileManagerImpl implements FileManager
+{
+    // The case of the FileManager used first.
+    private static Logger log = LoggerFactory.getLogger(FileManager.class) ;
+    
+    /** Delimiter between path entries : because URI scheme names use : we only allow ; */
+
+    static FileManager fmInstance = null ;
+
+    protected List<Locator> fmHandlers = new ArrayList<>() ;
+    protected LocationMapper fmMapper = null ;
+    
+    static { JenaSystem.init(); }
+
+    /** Get the global file manager.
+     * @return the global file manager
+     */
+    public static FileManager get()
+    {
+        // Singleton pattern adopted in case we later have several file managers.
+        if ( fmInstance == null )
```

Comment:
Should this method be made `synchronized`, or should we tell users of this code that it is not thread-safe? (I think it's not, but my Java-fu is getting utterly confused due to writing more JS & Python at $work)